### PR TITLE
Feature/shuffle reader

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: paddle/fluid/framework/reader.h [hunk 1]
```diff
@@ -16,40 +16,29 @@
...
 namespace framework {
 
 class ReaderBase {
  public:
-  explicit ReaderBase(const std::vector<DDim>& shapes) : shapes_(shapes) {
-    PADDLE_ENFORCE(!shapes_.empty());
-  }
   virtual void ReadNext(std::vector<LoDTensor>* out) = 0;
 
   virtual void ReInit() = 0;
```
[the remark this answers is not in the document]
Comment (reyoung) [reply paddle/fluid/framework/reader.h]: For a network reader, we can just throw an exception.

The reader module is in flux. I just discussed the design doc with @JiayiFeng  yesterday. Whole design doc will be written soon, maybe within 1-2 days.

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 2]
```diff
@@ -71,44 +107,73 @@ class CreateDoubleBufferReaderOpMaker : public DecoratedReaderMakerBase {
...
     )DOC");
+    std::unordered_set<std::string> enum_range;
+    constexpr size_t kMaxCUDADevs = 128;
+    for (size_t i = 0; i < kMaxCUDADevs; ++i) {
+      enum_range.insert(string::Sprintf("CUDA:%d", i));
+    }
+    enum_range.insert("CPU");
+    AddAttr<std::string>("place", "The double buffer place, default is CPU")
+        .SetDefault("CPU")
+        .InEnum({enum_range});
```
Comment (JiayiFeng) [inline paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: Shouldn't 'CPU' be in this enum?

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 3]
```diff
@@ -24,11 +24,31 @@ static constexpr size_t kDoubleBufferSize = 2;
...
-  explicit DoubleBufferReader(ReaderBase* reader)
-      : DecoratedReader(reader),
-        buffer_(framework::MakeChannel<std::vector<framework::LoDTensor>>(
-            kDoubleBufferSize)) {
-    std::thread prefetch(&DoubleBufferReader::PrefetchThreadFunc, this);
+  struct Item {
+    Item() : ctx_(nullptr) {}
+
+    std::vector<framework::LoDTensor> payloads_;
+    platform::DeviceContext* ctx_;
```
Comment (JiayiFeng) [inline paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: It seems that this `ctx_` will only hold gpu_ctx. So `gpu_ctx_` might be a better name.

File: paddle/fluid/operators/reader/create_double_buffer_reader_op.cc [hunk 4]
```diff
@@ -42,7 +62,10 @@ class DoubleBufferReader : public framework::DecoratedReader {
  private:
   void PrefetchThreadFunc();
 
-  framework::Channel<std::vector<framework::LoDTensor>>* buffer_;
+  framework::Channel<Item>* buffer_;
+  platform::Place place_;
+  std::vector<std::unique_ptr<platform::DeviceContext>> ctxs_;
```
Comment (JiayiFeng) [inline paddle/fluid/operators/reader/create_double_buffer_reader_op.cc]: The `DoubleBufferReader` only takes one `Place` and all ctx would be the same. So why do we need a vector here?

File: paddle/fluid/framework/reader.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (helinwang) [reply paddle/fluid/framework/reader.h]: Thanks for the prompt reply!

**Pull request merged**